pkg/stanza/operator/input/tcp/tcp.go

```diff
@@ -239,48 +245,75 @@ func (t *Input) goHandleMessages(ctx context.Context, conn net.Conn, cancel cont
 		defer t.wg.Done()
 		defer cancel()
 
+		if t.OneLogPerPacket {
+			var buf bytes.Buffer
+			io.Copy(&buf, conn)
+			log := truncateMaxLog(buf.Bytes(), t.MaxLogSize)
+			handleMessage(ctx, conn, t, log)
+			return
+		}
+
 		buf := make([]byte, 0, t.MaxLogSize)
+
 		scanner := bufio.NewScanner(conn)
 		scanner.Buffer(buf, t.MaxLogSize)
 
 		scanner.Split(t.splitFunc)
 
 		for scanner.Scan() {
-			decoded, err := t.encoding.Decode(scanner.Bytes())
-			if err != nil {
-				t.Errorw("Failed to decode data", zap.Error(err))
-				continue
-			}
+			handleMessage(ctx, conn, t, scanner.Bytes())
+		}
 
-			entry, err := t.NewEntry(string(decoded))
-			if err != nil {
-				t.Errorw("Failed to create entry", zap.Error(err))
-				continue
-			}
+		if err := scanner.Err(); err != nil {
+			t.Errorw("Scanner error", zap.Error(err))
+		}
+	}()
+}
 
-			if t.addAttributes {
-				entry.AddAttribute("net.transport", "IP.TCP")
-				if addr, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
-					ip := addr.IP.String()
-					entry.AddAttribute("net.peer.ip", ip)
-					entry.AddAttribute("net.peer.port", strconv.FormatInt(int64(addr.Port), 10))
-					entry.AddAttribute("net.peer.name", t.resolver.GetHostFromIP(ip))
-				}
+func handleMessage(ctx context.Context, conn net.Conn, t *Input, log []byte) {
+	decoded, err := t.encoding.Decode(log)
+	if err != nil {
+		t.Errorw("Failed to decode data", zap.Error(err))
+		return
+	}
 
-				if addr, ok := conn.LocalAddr().(*net.TCPAddr); ok {
-					ip := addr.IP.String()
-					entry.AddAttribute("net.host.ip", addr.IP.String())
-					entry.AddAttribute("net.host.port", strconv.FormatInt(int64(addr.Port), 10))
-					entry.AddAttribute("net.host.name", t.resolver.GetHostFromIP(ip))
-				}
-			}
+	entry, err := t.NewEntry(string(decoded))
+	if err != nil {
+		t.Errorw("Failed to create entry", zap.Error(err))
+		return
+	}
 
-			t.Write(ctx, entry)
+	if t.addAttributes {
+		entry.AddAttribute("net.transport", "IP.TCP")
+		if addr, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
+			ip := addr.IP.String()
+			entry.AddAttribute("net.peer.ip", ip)
+			entry.AddAttribute("net.peer.port", strconv.FormatInt(int64(addr.Port), 10))
+			entry.AddAttribute("net.peer.name", t.resolver.GetHostFromIP(ip))
 		}
-		if err := scanner.Err(); err != nil {
-			t.Errorw("Scanner error", zap.Error(err))
+
+		if addr, ok := conn.LocalAddr().(*net.TCPAddr); ok {
+			ip := addr.IP.String()
+			entry.AddAttribute("net.host.ip", addr.IP.String())
+			entry.AddAttribute("net.host.port", strconv.FormatInt(int64(addr.Port), 10))
+			entry.AddAttribute("net.host.name", t.resolver.GetHostFromIP(ip))
 		}
-	}()
+	}
+
+	t.Write(ctx, entry)
+}
+
+func truncateMaxLog(data []byte, maxLogSize int) (token []byte) {
+	dataLength := len(data)
+
+	if dataLength >= maxLogSize {
+		return data[:maxLogSize]
+	}
+
+	if dataLength == 0 {
```

Comment:
nit: I think calling len twice is probably preferable to the extra allocation.
```suggestion
	if len(data) > maxLogSize {
		return data[:maxLogSize]
	}
	if len(data) == 0 {
```